Support remote database access using external tools

t-h-e opened the pull request:
As discussed here: https://github.com/keepassxreboot/keepassxc/issues/1775#issuecomment-983056115

This change provides the ability to sync a database with a remote database with the help of `scp`. (Fixes #1775)
The `RemoteSettingsDialog` works similar to `DatabaseSettingsDialog` and can be extended in the future to store previous settings and load them again.
Although currently `scp` is supported, the implementation is extensible to include further programs in the future to download and upload .kdbx files to sync with them.

Feedback is very welcome as this would be my first contribution to KeePassXC

## Testing strategy
Place a second database on a separate server that is accesible via `ssh` and have `scp` installed and in your PATH variable.
The second database can include keys and/or groups that the current database does not.
The new "Remote Sync ..." feature is available in the UI next to "Merge From Database...".

Regarding unit tests: Help is very welcome as I am uncertain how to mock the QProcess. I expect the rest to work similarly to `TestGui::testMergeDatabase`

## Type of change
- ✅ New feature (change that adds functionality)

t-h-e commented:
@droidmonkey I think this PR is ready for review. As mentioned in the discussion of the issue, I'm not a C++ developer, so feedback is very welcome 😃

droidmonkey commented:
Excellent, will review when I get the chance

hifi commented:
I'll add a note that `scp` is considered harmful and this should probably use `sftp` instead.

t-h-e commented:
> I'll add a note that `scp` is considered harmful and this should probably use `sftp` instead.

While [OpenSSH recommends](https://www.openssh.com/txt/release-8.0) to use modern protocols like `sftp` or `rsync`, in the near future `scp` will be [running with SFTP protocol](https://www.openssh.com/txt/release-8.8). Since version 8.7, `scp` can already be used with [SFTP protocol with the switch `-s`](https://www.openssh.com/txt/release-8.7).
The benefit of `scp` over `sftp` is that a single file upload can be done with a single command, which makes it easier to use.

I have changes for using `rsync` and `curl` nearly finished, but I would prefer to at least get some feedback first to know if this feature is heading in the right direction. More protocols can easily be added.

hifi commented:
AFAIK you can use a single command with sftp if you batch it through stdin. That's without any temporary files.

I have high personal interest in this PR so I'll be testing this as well and providing feedback once I get around to.

hifi commented:
I took a quick look, it's promising but maybe it's not being utilized to its fullest potential with hardcoded commands.

I'm thinking sync over SSH is an advanced user feature so what if this would be simplified to "Sync Command" and instead provide a generic mechanism to run a command that can be templated with the database path and possibly other options? We'd only check that the return code of the command is a success and if not possibly show the code and stdout+stderr for debugging.

We could still provide presets for scp, sftp etc. for convenience. This allows the user to write script that takes the arguments from KeePassXC and does whatever with them.

KFDCompiled commented:
I've contributed to the #1775 bounty and am excited to test this PR!

t-h-e commented:
@hifi using templating and running arbitrary commands is a great idea. It is way more flexible. Running the sync command after unlocking as shown in [your comment](https://github.com/keepassxreboot/keepassxc/issues/1775#issuecomment-1051001958) would definitely be a convenient feature. 👍 
You seem to have way more keepassxc experience. I am not sure if I should even proceed with my PR then 😆

t-h-e commented:
@droidmonkey @phoerious Any chance to get a review for this PR? It is nearly 10 months old now 😞

phoerious commented:
Yes, sorry for the delay. But first, could you please rebase it to the latest develop HEAD and fix the conflicts?

t-h-e commented:
Done 😃 
Edit: I will check next week, why the build fails

t-h-e commented:
PR is rebased and conflicts are resolved. The build works mostly. The only issue seems to be that `TestGui::testSaveBackup` fails sometimes, but I cannot reproduce the behaviour locally.
Do you have any idea why this test might fail? It does not seem to fail for all build configurations.

droidmonkey commented:
I'll give this a review soon

t-h-e commented:
Hi, I added now the possibility to run any commend to download or upload a keepass file for syncing.
For some reason my test fails on the CI. I can execute the test on a windows and linux machine without issues though 🤷

HexF commented:
I've tested this out a little and come across a few quirks:
1. When doing a remote sync, the UI entirely locks up whilst downloading and uploading the database, giving no indication to the user as to what is going on.
2. While using the anyCommand feature with `rclone`, I managed to copy my Keepass database into a directory named with `{TEMP_DATABASE}`, rather than copying the file which produced no error up until I fill all the credentials in and get a "cannot open file" error.
3. When synchronizing the exact same database I am required to select my key file and enter my password again, this is a little inconvenient. I would much prefer the derived key from my currently open database be tried first before asking for my password and key file again.
4. Whenever I open the menu and click "Remote Sync", I am always brought to the Config UI, having to click OK before a sync will take place. I would prefer this only occur once.
5. If I restart KeepassXC, my I inputted into the config UI are lost and have to be re-entered. This appears to be indicated [here](https://github.com/keepassxreboot/keepassxc/pull/7222/files#diff-99961206e7bf329f7e9bc317bb061c465761ea4846da17f443c56e4d598e295eR49-R53)
6. The database name containing `{}` may cause some file name compatibility errors on some filesystems/operating systems.


I'll keep you updated as I use this feature more extensively, but for now its looking pretty good.

For those interested, here is the upload/download commands I'm using for syncing with Nextcloud:
```sh
# Download
rclone copyto nextcloud:Passwords.kdbx {TEMP_DATABASE}
# Upload
rclone copyto {TEMP_DATABASE} nextcloud:Passwords.kdbx
```

t-h-e commented:
@HexF Thank you so much for testing and your feedback. 😊 
Some comments from my side:
1. Yes, that should be handled better, but is further down the priority list. Although I think that can be fixed quickly
2. That is weird. Do you have more input on this error how to reproduce it?
3. This annoys me as well. The code is inspired by the merging feature, where this is also necessary. I have to check if there is a more convinient solution.
4. Hmm, that I understand, but what if you want to sync to a different location afterwards. Maybe the menu can have an option "sync with previous settings" or offer a drop down with the last options. This is probably connected to 5.
5. Not all functionality is implemented yet.
6. True, I might just use a dash.

On a side note: I had some time the last weeks and added an option to also be able to open a remote database. I will update the PR as soon as I have some tests and cleaned the code.
Afterwards I will work on your suggestions. Thanks again.

HexF commented:
> That is weird. Do you have more input on this error how to reproduce it?

I imagine using a command such as `mkdir {TEMP_DATABASE}` should do it.

> Hmm, that I understand, but what if you want to sync to a different location afterwards. Maybe the menu can have an option "sync with previous settings" or offer a drop down with the last options. This is probably connected to 5.

If multiple destinations are to be supported, I would prefer adding them in a preferences menu, then being able to select them in a flyout menu from the "Remote Sync" button. You could also put a quick-link to this preferences page in this flyout.

Something a little like this:
```
...
Remote Sync > | NextCloud
Import        | Google Drive
Export        | ---
Quit          | Edit remote sync destinations
```

t-h-e commented:
I have added a "remote open dialog" to allow people to open remote databases with KeePassXC directly. This uses a lot of the functionality that I added previously.
Next step is to address some of the issues mentioned by @HexF

HexF commented:
Looking at the remote open, I don't see any hook for a save to upload/merge the database with the remote one. Could this be added?

t-h-e commented:
Sure that is easy enough.
I still need to figure out what's wrong with the tests on TeamCity. Locally (windows and linux), all tests run fine.

mazunki commented:
I haven't tried out this feature yet, but I hope that when the feature is implemented, there would be a hook to automatically push passwords to `${DEFAULT_REMOTE}` whenever an update is done, and some mechanism to automatically pull from the remote each `${RESYNC_INTERVAL}` seconds. 

This would make it seamless to the user, provided that ssh is set up already.

Is this feasible to implement?

t-h-e commented:
Update:
* Remote sync does not require password input if the same key/password is used. If a different key/password is used, the user will be prompted to input as before
* Download/upload run concurrently and do not block GUI anymore
* Added more info in the status bar during remote sync
* Values (command, url, port etc.) used in the remote sync are stored in the config and are not lost anymore when restarting KeepassXC
* Braces in the filename have been removed

t-h-e commented:
@droidmonkey Can you help me figure out what is wrong with my tests? Locally everything works fine. I ran the tests on Windows 10 and Manjaro Linux. Nevertheless the pipeline fails consistently. As I cannot reproduce the failing tests, it is difficult to fix them. Can you reproduce the failing tests or have a look what is going on?

Also, which version of QT is the pipeline using? I had to remove the usage of `QUuid::WithoutBraces` as the symbol was not recognized. See commit https://github.com/keepassxreboot/keepassxc/pull/7222/commits/51ddff19efdf1ce9c17badf8e5a384bca2369144. This symbol was already introduced in [version 5.11](https://doc.qt.io/qt-6/quuid.html#StringFormat-enum)

droidmonkey commented:
You have a crash that is occurring on one of the download tests, specifically with the scp download

t-h-e commented:
Update:

- No more `scp` specific widget. All commands have to be defined in the same way (Previously named `anyCommand`)
- Multiple remote settings can be defined with a name
- Upload without sync is supported
- Menu offers the option to directly sync with a previously defined remote setting. The saved setting needs the "Add to Menu checked"

Remote sync settings:
![image](https://user-images.githubusercontent.com/999552/228668813-89a5d9d0-d982-496b-9e45-c98f902562e2.png)

Remote sync menu
![image](https://user-images.githubusercontent.com/999552/228668912-d29779d5-046f-480a-85ef-216d1cec4043.png)

@HexF Can you give it another look?
@mazunki For now I have decided against an automated sync mechanism. Mainly to get the current version out of the door. But yes, it can be done at some point.

@droidmonkey @phoerious Can one of you give the PR a review? I think the feature has matured since the first implementation 😃

droidmonkey commented:
This looks great! Will give a thorough review when I can, currently absolutely swamped

HexF commented:
I'll give this a test later tonight

HexF commented:
I'm a big fan of the changes made, especially the ability to save my upload/download scripts - makes all the difference.
Not having to enter my password & select my keyfile is also great.

Remote Opening appears to be a bit broken - it opens for me and pulls the latest database, but when I save & sync, the database doesn't update on the remote.

Remote opening also indicates the database path as `/tmp/RemoteDatabase-.....` in the open dialog, which isn't very user friendly:
![image](https://user-images.githubusercontent.com/11236933/229041166-1de52dee-aaff-4519-9339-9f5fd2856534.png)

My Recent Databases is also filled with these databases:
![image](https://user-images.githubusercontent.com/11236933/229041233-4d2830fa-44f5-4442-be45-b312d8fde5f7.png)

Ideally, these should say something along the lines of "NextCloud Remote Database" (replacing NextCloud with the name of the remote config)

When I remotely open I get a dialog like this after hitting OK with no feedback as to what is happening:
![image](https://user-images.githubusercontent.com/11236933/229041580-a3ec1adf-f343-49f0-a362-2373d951fffe.png)


I'm not sure that syncing is working at all for me, my commands are
```sh
# Download
rclone --config /run/secrets/thobson_rclone_config copyto nextcloud:Passwords.kdbx {TEMP_DATABASE}
# Upload
rclone --config /run/secrets/thobson_rclone_config copyto {TEMP_DATABASE} nextcloud:Passwords.kdbx
```
These commands to upload a file to nextcloud.

If I create an entry "Test Entry":
![image](https://user-images.githubusercontent.com/11236933/229042630-5f612ec6-5e10-4b39-81de-d8aaa80a8e8f.png)

then click Remote Sync -> NextCloud, I confirm on the bottom right action bar I see "Uploading"

I then download the file from NextCloud, opening it in Keepass.
I note the entry does not exist.

I am on commit 74220fa622989550ba54115a297d9bf45b5d35a1

My suspicion is this is to do with the actual merging step.

t-h-e commented:
@HexF Thanks for the detailed feedback.

- Bug: The uploaded file in case of both databases having the same key, has not been saved. Thanks for finding this one 😃
  - In case the keys are different, it works, but probably only due to the automatic save setting
- I will add a progress bar to the dialog so that the user can see that something is happening
- I forgot about the "Recent databases". Yes that can be improved. I will have a look, but don't see it as major blocker.

t-h-e commented:
- Bug fixed: Save source and destination database before uploading
- Remote databases are not added to recent databases
    - in the future it may make sense to add the remote settings recently used, but that is more complicated
    - using a name without UUID will lead to issues if opening a remote database and trying to sync it. Therefore I opted for suppressing remote database file paths for now
- Added status bar to remote file dialog

HexF commented:
I can confirm the bug has been fixed.

One thing I would like to see is being able to automatically sync my database when I lock it.

On the tabs up here, it would be nice to differentiate which one is open locally, and which one is open from a remote. At this stage its impossible to tell.
![image](https://user-images.githubusercontent.com/11236933/229944242-0baa25b9-3c9a-49c5-8c28-c596ea5265b3.png)

When I want to unlock the database, I am presented with the same `RemoteDatabase-` path, which in my mind should say something more like `Remotely Opened Database from NextCloud`

![image](https://user-images.githubusercontent.com/11236933/229944418-c7bc154d-4067-4363-a98c-2fb47edf00cc.png)

The documentation when setting up remote sources could be documented a bit better, especially the Command field, which doesn't indicate that you can use `{TEMP_DATABASE}` placeholder in it at all.

Other than those few things, I'm perfectly happy with the UX

t-h-e commented:
- added "[Remote]" in tab name
![image](https://user-images.githubusercontent.com/999552/232610109-4a1a9dbb-adb7-483f-9776-4277ba9f2a62.png)

- added another example for the remote dialog that `TEMP_DATABASE` can also be used in the command
![image](https://user-images.githubusercontent.com/999552/232610240-6fa6b01e-77e3-495e-b649-4b7fd803e250.png)

- remote databases are not saved in `LastDatabases`, `LastActiveDatabase` and `LastOpenedDatabases` as the paths saved may lead to non-existend temporary files.
- rebased the PR

As the feature has matured quite a bit and IMHO is ready to be reviewed/merged, I will refrain from further feature changes unless related to a review or bug fixes.

I'm looking forward to a review 😃 
(The failing MacOS test is also failing in other PRs)

dvilker commented:
Just fantasy out loud.

It could be done so that the program could "open" files like *.kdbx.sh.
And used it like this:
To read the file I ran mydb.kdbx.sh open and read the output.
And to write the file, I would run mydb.kdbx.sh save and write the file to the input.

Or read the first line #!... and run that. How shell does it.

And no UI would have to do.

droidmonkey commented:
You can use an external utility such as Watchman to do this. We will never execute external scripts from keepassxc.

KFDCompiled commented:
This works.  I started the bounty for #1775 and I think this satisfies all requirements.  Any objections to awarding bounty to @t-h-e ?

t-h-e commented:
@KFDCompiled I have submitted this PR to issuehunt. But maybe you should ask for objections in the [issue](https://github.com/keepassxreboot/keepassxc/issues/1775) as not as many might have subscribed for updates to the PR.

droidmonkey commented:
All I ask before I review and merge this... please go once over again and cleanup any temporary or messy code and add documentation.

t-h-e commented:
@droidmonkey I had another glance at the code and did a minor cleanup, otherwise the code looks well-structured to me. 
The PR has also been rebased again. Let me know if there is anything specific that needs to be changed or is blocking the merge from your point of view.

droidmonkey commented:
Ran through real quick, you are missing some `tr(...)` wrappers around text strings for gui display. You also need to add newlines to end of several files. Also get the documentation written up please.

t-h-e commented:
With documentation you mean a PR to https://github.com/keepassxreboot/keepassxc-org/blob/main/content/docs/KeePassXC_UserGuide.html ?
I have added the `tr(...)` wrappers and newlines.

droidmonkey commented:
No, either add a section on remote database operations to the below topic or create a new topic and link it in.

https://github.com/keepassxreboot/keepassxc/blob/develop/docs/topics/DatabaseOperations.adoc

These files compile into the html documentation

t-h-e commented:
Thanks, will do.
One more question. What tool did you use to annotate the screenshots with numbers, arrows etc.? I'm asking so that my screenshots have the same look

droidmonkey commented:
SnagIt, I can do screenshots if you like

droidmonkey commented on src/core/AsyncTask.h:
```diff
@@ -20,13 +20,13 @@
 
 #include <QFutureWatcher>
 #include <QtConcurrent>
+#include <utility>
```
This doesn't appear to be used

droidmonkey commented on src/core/AsyncTask.h:
```diff
@@ -54,6 +54,22 @@ namespace AsyncTask
+     *
+     * @param task std::function object to run
+     */
+    template <typename FunctionObject> void runAndWaitForFinished(FunctionObject func)
```
I don't prefer this to be added. You should use the non event loop blocking function that does a callback when done.

droidmonkey commented on src/core/Config.h:
```diff
@@ -194,6 +194,8 @@ class Config : public QObject
         Messages_Qt55CompatibilityWarning,
         Messages_HidePreReleaseWarning,
 
+        Remote_Program_Settings,
```
This doesn't follow naming patterns

t-h-e commented:
@droidmonkey I have added documentation for the remote database support. If you could replace the provisional screenshots with ones made with SnagIt so that they look more consistent and nicer that would be great. You may also want to update `docs/images/welcome_screen.png` as there is now a fourth button.

Let me know if there is anything else I need to add.

droidmonkey commented:
Awesome will review

droidmonkey commented:
OK here is a functionality review:

![image](https://github.com/keepassxreboot/keepassxc/assets/2809491/cba707ec-6801-44fe-ba49-2d4f258aff6b)

![image](https://github.com/keepassxreboot/keepassxc/assets/2809491/88c5afa7-0dde-48e6-bed9-f23323904bd8)

![image](https://github.com/keepassxreboot/keepassxc/assets/2809491/63fb008a-4599-403f-95b7-9fb83123c8f1)

Regarding the import option, I have also completely replaced the import menu with my upcoming Import Wizard PR. We can incorporate a remote file import into that wizard at a later time. 

I also do not like that these sync settings are stored in the configuration file. They should be stored in the Database Custom Data since they only apply to the specific database that is opened and could leak information (username/password) to the sync service. The "Open from remote" option should simply ask for an initial command string that opens the database and then writes that information to the sync data into that database for easy use to sync with later on.

t-h-e commented:
Thanks for the functional review.

I think most changes can easily be done.

Regarding the storage of the sync settings. They can be stored in `Database::pubicCustomData`, as each database may have their own sync settings. This would also allow to use placeholders as suggested here: https://github.com/keepassxreboot/keepassxc/issues/1775#issuecomment-1051001958
But I'm not sure this works well for "Open from remote" use case. If a remote database is opened, it is saved in a temporary folder. So unless the user saves it in another directory as well, it may be gone soon. If I would use "Open from remote", I may want to open the same remote database again the next time I open KeepassXC. Therefore, the settings need to be stored someplace other than within the database, so that as a user I can open the remote database again quickly without having to type the command each time.

Do you have a suggestion how to accommodate the "Open from remote" use case within KeepassXC? The only option I see is to store it within the config.

droidmonkey commented:
No, do not store in `publicCustomData` use  `customData`. The former is stored in the kdbx header and unencrypted. 

For the open remote database case, I guess you can keep those settings in the config. The user should be warned about storage of passwords in the command line. Alternatively, we could have another placeholder like {REMOTE_PASSWORD} which will prompt for the password on use.

QORTEC commented:
Why not give the option of caching the remote database (save in a "cache" folder)?
- If the database is cached, the credentials to sync can be retrieved from there. 
- If the database is not cached (/tmp), then we can assume that the user doesn't want the database easily accessible form the computer.


This would also enable us to give the option to open the cached database if the remote database is not accessible.
This would probably be read only, since merge/conflict logic required is definitely beyond the scope of this pull request.

Edits: minor punctuation/grammar fixes to clarify intent.

darktohka commented:
What is the status on this pull request? It would be nice to have remote database access.

t-h-e commented:
I hope to have some time next month to finish up the suggestions.
I haven't had much time or sleep (👶) the last couple of months

HexF commented:
Not sure if this is a bit late in the development cycle for this PR, or if I should just wait for this to be merged and PR it myself:

I would love to be able to use secrets from within the database in the commands.
I could see this being useful for syncing to WebDAV services such as NextCloud which require username/password or username/access token, which I already have stored in the database.

I was thinking similar to `{TEMP_DATABASE}` there can be a template such as `{54093383-9e73-42e6-b634-87fab8dfab81:password}` where `54093383-9e73-42e6-b634-87fab8dfab81` is the UUID of an entry in the database and `password` is the name of some attribute associated with the entry.

There are probably some security implications with this, such as the command being placed in publicly writable areas, which could lead to entry leakage by knowing a uuid of a record.
I'm sure there are ways to iron this out into a secure but user-friendly feature, such as requiring approval to access records in a similar way to FDO secrets and the browser extension.

Again, if you want to skip implementing this - no sweat, I'll get it done myself :)

t-h-e commented:
Finally I had a little bit of time.

- Remote settings are now part of database settings (see first screenshot)
- Settings are stored within `customData`
- Name cannot be an empty string anymore
- Open remote dialog has been reduced to the essential (see second screenshot)
  - Settings used for opening are _not_ stored anymore

![image](https://github.com/keepassxreboot/keepassxc/assets/999552/d6cdbef3-cc8e-402b-b30d-7df030176451)
![image](https://github.com/keepassxreboot/keepassxc/assets/999552/f08b31e1-86f8-482d-9f16-23e5259832d1)

@droidmonkey Can you have another look?

Documentation still needs to be updated.

MacOS tests currently break as I'm trying to open a menu to trigger the aboutToShow signal of QMenu. Freaking Mac 😡 
Not sure how to solve this or if I disable the remote tests.

t-h-e commented:
@HexF sounds great. In  a follow up PR 😉

droidmonkey commented:
Awesome will queue up another review.

t-h-e commented:
Updated remote sync documentation

I have also borrowed a Mac to check on the test issue. I setup the Mac and ran the tests. Everything worked. I rebased my changes and the pipeline now runs successfully for Mac as well. Seems that the issue was actually with the pipeline.
Now the pipeline fails in an unrelated test for Ubuntu. Not sure what is going on, but I guess that there are some flakey tests the break in the pipeline once in a while.

t-h-e commented:
Hi, @droidmonkey just pinging you regarding the review 😃

droidmonkey commented:
I made a bunch of edits to this PR I need to push, please don't rebase or merge anything for now

t-h-e commented:
Sure. Let me know if you need anything 😃

tw-atroehrsm commented:
meow? dont stop halfway through pls 🙏

droidmonkey commented:
I am done with my refactoring, testing the functionality now. There are a few areas that concern me with this capability in general, but it works! 

Feedback from the sync apps (sftp, etc) is generally hard to deal with. Even if a file download/upload fails when talking to sftp, it still returns a successful exit code because the app itself worked.

HexF commented:
Ideally in the documentation we should spell out that the exit code is responsible for determining success of a transfer.
Thus we should suggest people use `scp` over `sftp` and other application which require no stdin interaction/are consistent with exit code and transfers.

Maybe the entire stdin thing can be removed, which could be replaced by a call to a shell script on the user's system, providing much more flexibility for output parsing?

droidmonkey commented:
I have an idea, basically check that the temp database file actually exists and its size is not 0 bytes. Otherwise return an error that shows the stdout/stderr from the command.

t-h-e commented on src/gui/MainWindow.cpp:
```diff
@@ -703,6 +710,7 @@ MainWindow::MainWindow()
     m_progressBar->setMaximum(100);
     statusBar()->addPermanentWidget(m_progressBar);
     connect(clipboard(), SIGNAL(updateCountdown(int, QString)), this, SLOT(updateProgressBar(int, QString)));
+    connect(m_ui->tabWidget, SIGNAL(updateSyncProgress(int, QString)), this, SLOT(updateProgressBar(int, QString)));
```
This signal does not exist anymore.
`No such signal DatabaseTabWidget::updateSyncProgress(int, QString)`

t-h-e replied:
This error still occurs

t-h-e commented on src/gui/remote/RemoteProcess.cpp:
```diff
@@ -0,0 +1,76 @@
+    return m_process->readAllStandardOutput();
+}
+
+QString RemoteProcess::readError()
```
Good idea to have the error available

t-h-e commented on src/gui/remote/RemoteHandler.h:
```diff
@@ -0,0 +1,64 @@
+    explicit RemoteHandler(QObject* parent = nullptr);
+    ~RemoteHandler() override = default;
+
+    void sync(const QSharedPointer<Database>& db, RemoteParams* params);
```
`sync` is not implemented.

t-h-e commented on src/gui/remote/RemoteHandler.h:
```diff
@@ -0,0 +1,64 @@
+    };
+
+    // Used for testing only
+    static void setRemoteProcessFunc(std::function<QScopedPointer<RemoteProcess>(QObject*)> func);
```
`setRemoteProcessFunc` is not implemented anymore.
As the comment suggests, it was only used for testing, to be able to mock `RemoteProcess` as I did not want to start a real process during testing.

The method seems to be still used in test, which is one of the reasons why building the tests fails.

t-h-e commented on tests/gui/TestGui.cpp:
```diff
@@ -365,6 +370,107 @@ void TestGui::testMergeDatabase()
+    QString sourceToSync = "sftp user@server:Database.kdbx";
+    RemoteHandler::setRemoteProcessFunc([sourceToSync](QObject* parent) {
+        return QScopedPointer<RemoteProcess>(
+            new MockRemoteProcess(parent, QString(KEEPASSX_TEST_DATA_DIR).append("/SyncDatabase.kdbx"), sourceToSync));
```
`MockRemoteProcess` only takes a single argument after refactoring. This is why building tests fails at the moment.

t-h-e replied:
`MockRemoteProcess` only takes two arguments now.

t-h-e reviewed (commented):
Thanks for the refactoring.
The UI looks way better now, I love the new icon ❤️ and of course thanks for the code improvements.

I tested the changes on Windows and I'm running into the issue that the sync does not work anymore due to the following error:
```
QObject: Cannot create children for a parent that is in a different thread.
(Parent is RemoteHandler(0x1dfe3fa4720), parent's thread is QThread(0x1dfdc72c8d0), current thread is QThreadPoolThread(0x1dfe0e7d080)
```
I remember that I had this issue during development at some point in time as well, but cannot remember how I solved it.

I left a few comments regarding the code changes. The main issue is that the tests are not compiling at the moment, but I guess you are still working on the PR, right?

droidmonkey commented:
I need to push my latest, I'm slowly chipping away at the refactoring. It doesn't work perfectly right now, I'm refining user interface first then going back to backend to tighten everything up.

droidmonkey commented:
Take another look now

droidmonkey commented:
You can see my latest here: https://github.com/droidmonkey/keepassxc/tree/feature/remote-database-support

t-h-e commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1204,6 +1270,26 @@ void DatabaseWidget::mergeDatabase(bool accepted)
     emit databaseMerged(m_db);
 }
 
+bool DatabaseWidget::syncWithDatabase(const QSharedPointer<Database>& otherDb, QString& error)
```
👍 
It makes sense to only pass in `otherDb` rather than passing in two databases, as one can be accessed via `m_db`.
Also using `error` as reference rather than showing a message in this method is cleaner architecture

t-h-e commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1204,6 +1270,26 @@ void DatabaseWidget::mergeDatabase(bool accepted)
+bool DatabaseWidget::syncWithDatabase(const QSharedPointer<Database>& otherDb, QString& error)
+{
+    Merger firstMerge(m_db.data(), otherDb.data());
+    Merger secondMerge(otherDb.data(), m_db.data());
```
IMHO, I liked the previous naming better `mergerToRemote` and `mergerFromRemote`, rather than `firstMerge` and `secondMerge` as the new names to not convey the meaning why this merge is necessary.

droidmonkey replied:
I'm going to reuse this function for "sync with database" which is chosen from a file dialog. Nothing about this function is specific to a remote database since that aspect is handled from a different function.

t-h-e commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1035,6 +1043,64 @@ int DatabaseWidget::addChildWidget(QWidget* w)
+
+    auto result = AsyncTask::runAndWaitForFuture([this, params] {
+        QScopedPointer<RemoteHandler> remoteHandler(new RemoteHandler(this));
+        RemoteHandler::RemoteResult result;
```
I'm getting an error here, as the variable in line 1050 has the same name is therefore shadowed

t-h-e commented on tests/gui/TestGui.cpp:
```diff
@@ -367,6 +372,107 @@ void TestGui::testMergeDatabase()
+        return QScopedPointer<RemoteProcess>(
+            new MockRemoteProcess(parent, QString(KEEPASSX_TEST_DATA_DIR).append("/SyncDatabase.kdbx"), sourceToSync));
+    });
+    QSignalSpy dbSyncSpy(m_dbWidget.data(), &DatabaseWidget::databaseSyncedWith);
```
`databaseSyncCompleted` instead of `databaseSyncedWith`?
Same in test `testRemoteSyncDatabaseRequiresPassword`

t-h-e reviewed (commented):
The architecture of the code is way nicer than before. Thanks for cleaning up the code 👍 

The sync works again with the newest changes 😃

The "Open Remote Database" still has a bug though, but I assume you know it already.
After clicking OK, with any command, the progress bar appears, but the dialog buttons are still clickable.
Nothing happens after clicking OK. Neither if the download command should work or fail.

t-h-e committed:
Provide remote database sync capability

Allow any command to be executed for syncing databases

tburny commented:
Please continue working on this, I am eagerly waiting to move away from KeePass2 on Ubuntu and this is the only feature that prevents me from doing so.

Please let everyone know if you need any help. I'd be more  than happy even to sponsor this project once this feature works well.

droidmonkey commented:
Now that 2.7.7 is out I will resume this work

KaKi87 commented:
Hi,
How is it going ?
Thanks

roelbroersma commented:
@droidmonkey  and  @t-h-e  Keep up the good work! 👍 👍  I know it can be hard and bureaucratic to pass the whole PR process, but once it's past,.. you did it!  Come on, the last few inches..!

t-h-e commented:
Hi @droidmonkey, is there anything I can help with to get this PR over the finish line?

droidmonkey commented:
I lost my focus on this one, can you just look through the code and clean it up. I just pushed my latest commit without testing it, looks like I still need to remove some legacy artifacts from all the other code changes we have been doing.

```
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:255:36: error: out-of-line definition of 'importFile' does not match any declaration in 'DatabaseTabWidget'
[00:13:14] :	 [Step 2/4] DatabaseWidget* DatabaseTabWidget::importFile()
[00:13:14] :	 [Step 2/4]                                    ^~~~~~~~~~
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:343:25: error: out-of-line definition of 'importKeePass1Database' does not match any declaration in 'DatabaseTabWidget'
[00:13:14] :	 [Step 2/4] void DatabaseTabWidget::importKeePass1Database()
[00:13:14] :	 [Step 2/4]                         ^~~~~~~~~~~~~~~~~~~~~~
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:357:15: error: no member named 'switchToImportKeepass1' in 'DatabaseWidget'
[00:13:14] :	 [Step 2/4]     dbWidget->switchToImportKeepass1(fileName);
[00:13:14] :	 [Step 2/4]     ~~~~~~~~  ^
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:360:25: error: out-of-line definition of 'importOpVaultDatabase' does not match any declaration in 'DatabaseTabWidget'
[00:13:14] :	 [Step 2/4] void DatabaseTabWidget::importOpVaultDatabase()
[00:13:14] :	 [Step 2/4]                         ^~~~~~~~~~~~~~~~~~~~~
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:378:15: error: no member named 'switchToImportOpVault' in 'DatabaseWidget'
[00:13:14] :	 [Step 2/4]     dbWidget->switchToImportOpVault(fileName);
[00:13:14] :	 [Step 2/4]     ~~~~~~~~  ^
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:676:9: error: use of undeclared identifier 'db'
[00:13:14] :	 [Step 2/4]     if (db->isRemoteDatabase()) {
[00:13:14] :	 [Step 2/4]         ^
[00:13:14] :	 [Step 2/4] /Users/KPXC/buildAgent/work/c401303cba1b4098/src/gui/DatabaseTabWidget.cpp:681:9: error: use of undeclared identifier 'db'
[00:13:14] :	 [Step 2/4]     if (db->isModified()) {
[00:13:14] :	 [Step 2/4]         ^
[00:13:14] :	 [Step 2/4] 7 errors generated.
```

t-h-e commented:
@droidmonkey If I remember correctly, you suggested to integrate opening a remote database into the import wizard, right?
Here is a mockup:
![image](https://github.com/keepassxreboot/keepassxc/assets/999552/1d46de32-1ed9-43e8-b47c-4b3f1fc8dcd3)

I have added the "Command" and "Input" fields to download the remote database file.
I'm not sure about with options should be available for the "Import Into" section. I have added an option "Temporary Database" in the mockup, which would open the database, but not store it anywhere, except in the `tmp` folder of your OS, which will get deleted at some point. Another possible option would be to specify a save location, like "Save Database to:"

Do you have any suggestions?

t-h-e committed:
get current version compiling

droidmonkey commented:
Hey that's nifty, I don't recall making that suggestion though. New database and existing database seem accurate to me. Temporary database doesn't make much sense if I'm going through all this trouble to specify the remote settings. Basically you want to land the user with an "in memory" database that needs to be saved prior to close which will force the file choice dialog. That is what New Database generally does, but in this case you wouldn't need to show the new database workflow.

t-h-e committed:
Remove open remote database dialog

opening a remote database will be integrated into the import dialog

michaelk83 commented:
> integrate opening a remote database into the import wizard

The mockup looks overwhelming at first glance. What do you think about moving the command and input fields to the next wizard page? So you'd have the normal import wizard on the 1st page, then if you selected remote database, you're basically given a view of the remote access settings page, populated with your default settings. Then you can review your settings and adjust them just for the current import.

And I think this might be best to leave to a separate PR.

t-h-e committed:
Fix remote sync

- fix opening remote database with different key
- upload temporary database as local database may have different e.g. key
- fix tests
- further cleanup of open remote database

t-h-e committed:
update translations

t-h-e commented:
@michaelk83 I would refrain from adding another page in the wizard as it can get annoying to click through many pages in a wizard. The fields for downloading the database are only shown in case "Remote Database" is selected anyway. But I agree with the suggestion of having the import/open of a remote database in a separate PR. I should have done that in the first place.

@droidmonkey The code is compiling again.
I have added the functionality again that a remote database with a different key can be unlocked and synced.
The tests are working again (hopefully the CI agrees)
The progress bar that I added at some point seems to be gone. The progress bar displayed if the application is currently downloading, syncing or uploading so that the user is aware the something is going on rather than thinking the application is not responding. Was the progress bar removed by accident in one of the refactorings or was it on purpose? In the first case, I can add it again.

Otherwise, I think the PR is ready for review.

PS: I have left out the import wizard for now and will link a follow up branch/PR in the next couple of days that adds support for opening/importing a remote database. Is that alright?

droidmonkey commented:
Yup I agree, thank you for getting this back on track.

t-h-e committed:
Show progress bar when syncing

t-h-e commented:
@droidmonkey I have solved some merge conflicts and added the progress bar again.

Additionally, here is the PR for the support of opening/importing a remote database: https://github.com/t-h-e/keepassxc/pull/1
Currently the "open remote database" PR is only in my fork as it is based on this open PR to be merged first.

Let me know if you have any feedback, so that we can get this PR merged.

winnieXY commented:
Any news on that?

t-h-e commented:
@droidmonkey Is there any chance to get a review any time soon?

droidmonkey commented on src/core/CustomData.cpp:
```diff
@@ -27,6 +27,7 @@ const QString CustomData::BrowserLegacyKeyPrefix = QStringLiteral("Public Key: "
 const QString CustomData::ExcludeFromReportsLegacy = QStringLiteral("KnownBad");
 const QString CustomData::FdoSecretsExposedGroup = QStringLiteral("FDO_SECRETS_EXPOSED_GROUP");
 const QString CustomData::RandomSlug = QStringLiteral("KPXC_RANDOM_SLUG");
+const QString CustomData::RemoteProgramSettings = QStringLiteral("Remote_Program_Settings");
```
I really want to name this `KPXC_REMOTE_SYNC_SETTINGS`, we should do this now but it'll break anyone who used this PR before the change.

t-h-e replied:
Renamed

droidmonkey commented on src/core/Database.cpp:
```diff
@@ -973,7 +973,7 @@ void Database::markAsClean()
     stopModifiedTimer();
     m_hasNonDataChange = false;
-    if (emitSignal) {
+    if (emitSignal && !m_isRemoteDatabase) {
```
I really don't like this. We should always emit signals when state changes. You can (and should) choose to ignore them as the receiver based on your criteria. So if we want to ignore the fact that a database is "clean" if it is remote, then we should do that at the tab widget or database widget level. 

However this variable is also named wrong. In reality this is `m_isTemporaryDatabase` based on its use.

t-h-e replied:
True. I removed the condition here. I don't think it is needed anymore anyway.
Also renamed the variable and corresponding methods.

droidmonkey commented on src/gui/DatabaseOpenDialog.cpp:
```diff
@@ -201,6 +201,9 @@ void DatabaseOpenDialog::complete(bool accepted)
     m_db = m_view->database();
+    if (m_db != nullptr && m_intent == Intent::RemoteSync) {
+        m_db->markAsRemoteDatabase();
+    }
```
This is a weak association between the unlock dialog and the database being "remote". Technically we are marking this because the database is temporary in nature. This should be done at the point this distinction is made and not the open dialog. The open dialog should be left "neutral" to the type of database it is opening.

Unfortunately our unlocking logic is very complicated. It goes something like this:
DatabaseOpenDialog -> dialogFinished -> DatabaseWidget -> unlockDatabase -> syncUnlockedDatabase -> ....

DatabaseOpenDialog -> dialogFinished -> DatabaseTabWidget -> handleDatabaseUnlockDialogFinished

I'm not really sure what to do here but the complexity is a problem.

t-h-e replied:
This complexity really took me some time to understand. I haven't found a better way to handle this 😔

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -447,9 +451,13 @@ void DatabaseWidget::replaceDatabase(QSharedPointer<Database> db)
     m_db = std::move(db);
+    if (oldDb->isRemoteDatabase()) {
+        m_db->markAsRemoteDatabase();
+    }
```
Not sure if this holds true based on the usage of this variable. The old database being "remote" doesn't mean the new database is "remote" (or temporary?).

t-h-e replied:
True. I'm not sure anymore why I needed to add this. I removed this piece of code and will recheck it, before creating a PR for opening remote databases (https://github.com/t-h-e/keepassxc/pull/1)

droidmonkey reviewed (changes requested):
General Notes:
* Locking a database while attempting to sync keeps the UI disabled. Should cancel the sync immediately or wait until it is finished. I prefer the latter.
* There needs to be a "Test" button on the download sync settings page. Its a frustrating workflow to set the settings, save, OK, Database -> Remote Sync -> [Sync Name]. Obviously cannot test upload since that would cause data loss. Download can just run through the sequence and discard any resulting file.
* I cannot get this to work on Windows for whatever reason. It used to work but not any more. I suspect we may need to use this function to manipulate the underlying CreateProcess call: https://doc.qt.io/qt-5/qprocess.html#CreateProcessArgumentModifier-typedef

droidmonkey commented:
Pushing some initial feedback

droidmonkey committed:
Minor cleanup

t-h-e committed:
Fix messages in RemoteHandler

t-h-e committed:
Add test download button

t-h-e committed:
Disable Menu and Toolbar on Sync

E.g. locking the database while syncing can leave the UI in a
 non-functional state

t-h-e committed:
Rename m_isRemoteDatabase to m_isTemporaryDatabase

Remove m_isRemoteDatabase check in markAsClean as it serves no purpose

t-h-e committed:
Properly handle cancelled unlock dialog during remote sync

t-h-e committed:
Remove unused marking of temporary database

t-h-e committed:
update translations

t-h-e committed:
Rename remote program settings key

t-h-e commented:
- ad "Locking a database": The menu bar and toolbar get disabled now during syncing. This helps avoid issues with all the other functionality available in the menu and toolbar as well.
- added a "Test" button for testing the download
- ad Windows: I'm using Windows 11 Pro 23H2 and everything works fine.

droidmonkey commented:
I think my problems are with sftp, using password to unlock ssh key doesn't work.

t-h-e commented:
> I think my problems are with sftp, using password to unlock ssh key doesn't work.

Yes, that does not work. I run into the same issue in the beginning

droidmonkey commented:
Let's note that in the documentation. Workaround is to load the key to ssh-agent using the SSH Agent plugin.

t-h-e committed:
Add warning regarding password prompts for remote sync in documentation

t-h-e commented:
I have added a comment in the documentation.
All review comments have been addressed. Thanks for reviewing :smiley: 
Let me know if everything looks good now.

droidmonkey committed:
Minor cleanup

droidmonkey committed:
Final cleanup

droidmonkey commented:
OK this is ready to merge once CI finishes! Great work. I added a check for changes in the remote settings page so you don't lose work by accident. Also some general cleanup.

varjolintu commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1074,6 +1079,87 @@ int DatabaseWidget::addChildWidget(QWidget* w)
+    result.errorMessage = tr("Remote Sync did not contain any download or upload commands.");
+
+    // Download the database
+    if (!params->downloadCommand.isEmpty()) {
```
Even if I have configured both Upload and Download commands, even the Upload command goes inside this function every time and fails. What am I doing wrong?

EDIT: Starting things from the scratch worked. No idea why it didn't save the upload command at all.

KaKi87 commented:
Hi,
Would it be possible to have a beta build for this ?
Thanks

droidmonkey commented:
https://snapshot.keepassxc.org

KaKi87 commented:
That's nightly, not beta :sweat_smile:

droidmonkey commented:
That is our beta...

KaKi87 commented:
Well, I was hoping for a build that contains everything stable + remote databases support in addition, so that other experimental things wouldn't potentially add bugs to it. :/

segln commented:
Then how about just waiting for the next stable build?

KaKi87 commented:
This PR is almost 3 years old, the milestone is only 32% complete containing many other posts that are years old, the latest minor version increment was done 4 years ago, all signs indicating that this PR will not be released in a stable build anytime soon...

droidmonkey commented:
2.7.9, the stable version, is 95-98% the same as a snapshot build off develop branch. Just use a snapshot and don't think too hard about it. Trust me you'll be OK.

KaKi87 commented:
What's with the very worrying warning message then ? The threat of loosing passwords is a big deal.

droidmonkey commented:
We are toning that down, its just to keep people from using non-final builds in production.

Maasea commented:
I tried the beta version and implemented data synchronization for webdav using the curl command, which is easy to use. 

But the data synchronization is manual, is it possible to add a switch to automatically trigger remote synchronization for data changes, and to automatically pull remote files for updates every time the database is unlocked.

KaKi87 commented:
> its just to keep people from using non-final builds in production

But that's what I would be doing 😅

droidmonkey commented:
@Maasea that can be built in, maybe on a periodic basis versus every save to avoid locking up the interface. 

@KaKi87 if you are just using it yourself should be fine, don't deploy to your enterprise though. I am running the 2.7.10 beta build myself to ensure no bugs crept in

alexp303 commented:
Any idea when will the sync be included in the stable release?

droidmonkey commented:
This still needs a bit of trim work for UX, but you can use it now in our snapshot builds. They are rather stable, every change is well vetted. https://snapshot.keepassxc.org

wschlich commented:
> I tried the beta version and implemented data synchronization for webdav using the curl command, which is easy to use.
> 
> But the data synchronization is manual, is it possible to add a switch to automatically trigger remote synchronization for data changes, and to automatically pull remote files for updates every time the database is unlocked.

Can you please share your WebDAV sync implementation for KeePassXC using curl? Thank you!

wschlich commented:
@droidmonkey The current solution using remote sync settings per database stored inside the database itself (I guess in the KDBX headers?) is an interesting idea, I successfully tried using it with WebDAV and 2 cURL commands.

But how should it work using AutoOpen for other WebDAV DBs as supported by both (!) KeePass and KeePass2Android?

Right now, I have a main KDBX file on a WebDAV server.
I can just open the HTTPS-URL using KeePass and KeePass2Android and specify I/O username + I/O password for WebDAV (these can be cached by both KeePass and KeePass2Android so I don't need to enter them every time when I open that main DB).

In that main KDBX file, there's an AutoOpen folder, containing entries for other KDBX databases on other WebDAV servers.
These entries have an HTTPS-URL as well as IocUserName + IocPassword String fields which magically allow KeePass to open these WebDAV DBs without needing any more information.

Something like this (plain easily handleable!) seems to be impossible with the current DB-internal remote sync solution of KeePassXC? :-(

droidmonkey commented:
That is not supported and likely won't be. For awareness, keepass and keepass2android use the .NET framework which has support for web dav natively. Qt does not have this support and we are not interested in adding a third party library to add it. 

You could maybe achieve this in keepassxc by mounting the webdav folder in your OS and pointing your auto open to the mounted folder.

wschlich commented:
@droidmonkey 

> That is not supported and likely won't be. For awareness, keepass and keepass2android use the .NET framework which has support for web dav natively. Qt does not have this support and we are not interested in adding a third party library to add it.

Too bad, as there's a library available: https://github.com/PikachuHy/QtWebDAV

> You could maybe achieve this in keepassxc by mounting the webdav folder in your OS and pointing your auto open to the mounted folder.

Not portable across KeePass and KeePass2Android :-(

What about this idea:

As you already store DB remote sync settings inside the KDBX, what about adding support for the exact same kind of settings per AutoOpen entry as well, using String Fields, just like KeePass + KeePass2Android do with IocUserName + IocPassword, while ignoring the AutoOpen HTTPS-URL, but just using it as a {TEMP_DATABASE}?

To me, my master copy is the WebDAV remote anyway and not any local temporary/cached copy (as KeePassXC now seems to handle it).

droidmonkey commented:
You can have as many auto open entries as you want and use IfDevice. https://keepassxc.org/docs/KeePassXC_UserGuide#_automatic_database_opening

Extending this feature is likely not going to happen, it's niche to begin with and distraction from other important core features.

wschlich commented:
@droidmonkey
I believe you're completely missing my point here. It's about automatically opening *remote* WebDAV DBs using AutoOpen entries.
Please, read my comments again :-)

wschlich commented:
@droidmonkey 
The current implementation of remote database support in KeePassXC kills my highly portable workflow with multiple WebDAV DBs (one master and several AutoOpen WebDAV DBs specified inside) across KeePass2Android and classic KeePass -- it's really sad as I'd *love* to switch to KeePassXC on my Linux box because of .NET woes with mono on Linux :-(

droidmonkey commented:
I'm not missing your point. I'm telling you we aren't adding more features to this aspect of the program.

wschlich commented:
Very sad to read this. Thanks anyway :-)

mazunki commented:
> This still needs a bit of trim work for UX, but you can use it now in our snapshot builds. They are rather stable, every change is well vetted. https://snapshot.keepassxc.org/

Forgive my confusion, but does this mean it will be included in the 2.8.0 release? I suspect there's no specific ETA for that yet?

droidmonkey commented:
Yes and no eta right now.

kschmid commented:
Just tried out the remote sync feature from the most recent snapshot. This looks cool. Together with curl, I was able to successfully set up WebDav transfer. I am only confused about two things. 
a) The way I understand it, there is no way to have the local file at a fixed position? This confuses me. When I am traveling and have no remote access, it would be better to be able to access the most recent version at a local point. 
b) it seems, the only time remote sync happens, is when explicitly called for. - While this is nice, it might be even less errorprone, if remote-sync would be connected to opening dbs and (auto-)saving them. 

I hope, I got everything right. I tried to locate a corresponding user guide version, but it seems, there is not yet relevant doc for this.

droidmonkey commented:
Your local file is wherever your kdbx file is stored. Sync pulls down a temporary remote copy, performs a merge with the open copy, then saves the result locally. 

Remote sync is explicit because it's not meant to be "remote access".

kschmid commented:
Thanks a lot for the clarification with the merging. This escaped me. Just tried it out: very nice. 
Then also the explicit sync is not as problematic. - Though would still prefer "automatic merging", but if this is the decision...
Took me time to understand that these settings are also encrypted, so passwords could be part of the commands (which would simplify a lot. Overall very nice feature; looking forward to general release! Thanks for the great work!

droidmonkey commented:
Thank you for the feedback, we should make these things clearer
